postgis/lwgeom_in_marc21.c

```diff
@@ -0,0 +1,527 @@
+
+		if (literal_length < 4) {
+			POSTGIS_DEBUGF(3, "  invalid literal length (%d): \"%s\"", literal_length, literal);
+			return LW_FALSE;
+		}
+
+		coord_start = 1;
+	}
+
+	for (j = coord_start; j < literal_length; j++) {
+
+		if (!isdigit(literal[j])) {
+
+
+			if (j < 3) {
+
+				/**
+				 * The first three characters represent the degrees and they must be numeric.
+				 */
+
+				POSTGIS_DEBUGF(3,"  invalid character '%c' at the degrees section: \"%s\"",	literal[j], literal);
+				return LW_FALSE;
+
+			}
+
+			/**
+			 * Decimal degrees are encoded with after the 3rd character with either a dot or a comma.
+			 * Only one decimal separator is allowed.
+			 */
+			if (literal[j] == '.' || literal[j] == ',') {
+
+				num_dec_sep++;
+
+				if (num_dec_sep > 1) return LW_FALSE;
+
+			} else {
+				POSTGIS_DEBUGF(3, "  invalid character '%c' in %d: \"%s\"",	literal[j], j, literal);
+				return LW_FALSE;
+
+			}
+
+		}
+
+	}
+
+	POSTGIS_DEBUGF(2, "=> is_literal_valid returns LW_TRUE for \"%s\"",	literal);
+	return LW_TRUE;
+
+}
+
+static double parse_geo_literal(char *literal) {
+
+	/**
+	 * Coordinate formats supported (from the MARC21/XML documentation):
+	 *
+	 *  -> hdddmmss (hemisphere-degrees-minutes-seconds)
+	 *  -> hddd.dddddd (hemisphere-degrees.decimal degrees)
+	 *  -> +-ddd.dddddd (hemisphere[+/-]-degrees.decimal degrees)
+	 *     (“+” for N and E, “-“ for S and W; the plus sign is optional)
+	 *  -> hdddmm.mmmm (hemisphere-degrees-minutes.decimal minutes):
+	 *  -> hdddmmss.sss (hemisphere-degrees-minutes-seconds.decimal seconds)
+	 *
+	 */
+
+	char *dgr;
+	char *min;
+	char *sec;
+	char *dec;
+	char *csl;
+
+	int literal_length;
+
+	char hemisphere_sign = literal[0];
+	int start_literal = 0;
+	double result = 0.0;
+
+	POSTGIS_DEBUGF(2, "parse_geo_literal called (%s)", literal);
+	POSTGIS_DEBUGF(2, "  start character: %c", hemisphere_sign);
+
+	literal_length = strlen(literal);
+
+	if (!isdigit(hemisphere_sign)) {
+
+		start_literal = 1;
+	}
+
+	POSTGIS_DEBUGF(2, "    var start_literal=%d", start_literal);
+
+	dgr = malloc(literal_length);
+	memcpy(dgr, &literal[start_literal], 3);
+	dgr[3]='\0';
+
+	if (strchr(literal, '.') == NULL && strchr(literal, ',') == NULL) {
+
+		/**
+		 * degrees/minutes/seconds: hdddmmss (hemisphere-degrees-minutes-seconds)
+		 * Ex. +0793235
+		 *     E0793235
+		 *     ||  | |_seconds
+		 *     ||  |_minutes
+		 *     ||_degrees
+		 *     |_start_literal (hemisphere)
+		 */
+
+		POSTGIS_DEBUG(2, "  lat/lon literal detected");
+		POSTGIS_DEBUGF(2, "  degrees: %s", dgr);
+
+		POSTGIS_DEBUGF(5, "  min = malloc(%d)", literal_length-5);
+		min = malloc(literal_length - 5);
+
+		if (literal_length > (start_literal + 3)) {
+
+			memcpy(min, &literal[start_literal + 3], 2);
+			min[2] = '\0';
+		}
+
+		POSTGIS_DEBUGF(2, "  lat/lon minutes: %s", min);
+		POSTGIS_DEBUGF(5, "  sec = malloc(%d)", literal_length-5);
+		sec = malloc(literal_length - 5);
```

Comment:
palloc not malloc in the postgis/ directory

Reply:
malloc calls changed in #[fa23e6a](https://github.com/postgis/postgis/pull/659/commits/fa23e6aaaa14d03b0f515de1e44b5df2315eac95)